src/normalizer/DecompFile.zig

```diff
@@ -133,14 +246,132 @@ pub fn compressToFile(self: *DecompFile, filename: []const u8) !void {
 pub fn compressTo(self: *DecompFile, writer: anytype) !void {
     var buf_writer = std.io.bufferedWriter(writer);
     var out = std.io.bitWriter(.Little, buf_writer.writer());
-    try out.writeBits(@intCast(u16, self.entries.items.len), 16);
+
+    // For the UDDC registers, we want one register to represent each possible value in a single
+    // entry; we will emit opcodes to modify these registers into the desired form to produce a
+    // real entry.
+    var registers = std.mem.zeroes(Entry);
+
     while (self.next()) |entry| {
-        try out.writeBits(entry.key_len, 3);
-        _ = try out.write(entry.key[0..entry.key_len]);
-        try out.writeBits(@enumToInt(entry.value.form), @bitSizeOf(Form));
-        try out.writeBits(entry.value.len, 5);
-        for (entry.value.seq[0..entry.value.len]) |s| try out.writeBits(s, 21);
+        // Determine what has changed between this entry and the current registers' state.
+        var diff_value_form: u2 = undefined;
+        const diff = entry.diff(registers, &diff_value_form);
+
+        // If you want to analyze the difference between entries, uncomment the following:
+        //std.debug.print("diff={}\n", .{diff});
+        //continue;
+
+        // Infrequently changed: key_len, value.form, and value.len registers. Emit opcodes to
+        // update them if needed.
+        if (diff.key_len != 0) {
+            try out.writeBits(@enumToInt(Opcode.set_key_len_and_key), @bitSizeOf(Opcode));
+            try out.writeBits(entry.key_len, 3);
+            _ = try out.write(entry.key[0..entry.key_len]);
+        }
+        if (diff_value_form != 0) {
+            try out.writeBits(@enumToInt(Opcode.set_value_form), @bitSizeOf(Opcode));
+            try out.writeBits(diff_value_form, @bitSizeOf(Form));
+        }
+        if (diff.value.len != 0) {
+            try out.writeBits(@enumToInt(Opcode.set_value_len), @bitSizeOf(Opcode));
+            try out.writeBits(entry.value.len, 5);
+        }
+
+        // Frequently changed: this is where the magic happens.
```

Comment:
About this section: I want to say that I think we can probably improve readability here in a few ways. If you squint closely at it, you'll start to notice that actually most of these sections are quite simple - but at first glance the number of conditionals can seem a bit daunting. That's one thing I'd like to see if I can improve here.